pkg/etcd/etcdproxy.go

```diff
@@ -0,0 +1,76 @@
+package etcd
+
+import (
+	"net/url"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/k3s/pkg/agent/loadbalancer"
+)
+
+type Proxy interface {
+	Update(addresses []string)
+	ETCDURL() string
+	ETCDAddresses() []string
+	ETCDServerURL() string
+}
+
+// NewETCDProxy initializes a new proxy structure that contain a load balancer
+// which listens on port 2379 and proxy between etcd cluster members
+func NewETCDProxy(enabled bool, dataDir, etcdURL string) (Proxy, error) {
+	e := &etcdproxy{
+		dataDir:        dataDir,
+		initialETCDURL: etcdURL,
+		etcdURL:        etcdURL,
+	}
+
+	if enabled {
+		lb, err := loadbalancer.New(dataDir, loadbalancer.ETCDServerServiceName, etcdURL, 2379)
+		if err != nil {
+			return nil, err
+		}
+		e.etcdLB = lb
+		e.etcdLBURL = lb.LoadBalancerServerURL()
+	}
+
+	u, err := url.Parse(e.initialETCDURL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse %s", e.initialETCDURL)
```

Comment:
You don't need to print the URL that failed, the error from url.Parse will print it for you:
```
parse "http://192.168.0.%31:8080/": invalid URL escape "%31"
```
```suggestion
		return nil, errors.Wrap(err, "failed to parse etcd client URL")
```

Reply:
fixed